### PR TITLE
Support component mounting and unmounting in Turbolinks and PJAX

### LINKED ISSUE #13: Support for Turbolinks
Another gem: react-rails implements react-ujs:

https://github.com/reactjs/react-rails/tree/master/lib/assets/javascripts

maybe we could steal it to provide the same experience for turbolinks?

### PR DESCRIPTION
**Pull request by sevos**
React components will be mounted and unmounted on events specific for:

* Turbolinks 5
* Turbolinks 2.4+
* old, deprecated Turbolinks
* PJAX
* without any partial page loading technology present

This code is copied with modifications from https://github.com/reactjs/react-rails with modifications.

This resolves #13

### CODE REVIEW

**Comment by sevos** [comment]
TBH: I've tested it only on turbolinks 5 and native.

**Comment by sevos** [comment]
We need some test coverage for the new features

**Comment by sevos** [comment]
There are still some glitches when visiting page multiple times via turbolinks:

```
Warning: unmountComponentAtNode(): The node you're attempting to unmount was rendered by another copy of React.
```

**Comment by sevos** [comment]
React rails has the same issue https://github.com/reactjs/react-rails/issues/607

**Comment by renchap** [comment]
Can you fork https://github.com/renchap/webpacker-react-example and share a branch showing with this behaviour?

From https://github.com/facebook/react/pull/8928#issuecomment-277726453 I think Turbolinks reloads the pack file on visit, hence reloading React.

**Comment by sevos** [comment]
It's just when we use turbolinks5 and unmount components on `turbolinks:before-render`, then as described in the ticket above, the warning happens. A quick solution would be to use `turbolinks:before-cache` to unmount the component. I know, the component won't get cached in turbolinks that way, but if this gets rid of these warnings, I would be fine.

**Comment by renchap** [comment]
As gaeron said in the comment I linked, it does not explain why React is complaining about "another version of React". This seems to be the real issue to fix, rather than a workaround.
I will try to reproduce it and have a look.

**Comment by sevos** [comment]
To reproduce, you take your example app, you `yarn link` my branch of webpacker-react and launch everything. Then you run `Turbolinks.visit(window.location.href)` twice in the js console.

**Comment by sevos** [comment]
I think, I figured this out. I added few debugs and this is what I saw before the fix:

screen-shot-2017-02-25-18-31-30

On subsequent navigation `before-render` is called twice: once for cached page (with orphaned components) and once for the content loaded from the server. The warning happens when we try to unmount components loaded from the cache. 

With the changes, we ensure that components are mounted after loading from the cache. This will ensure that they are still functioning, even when connection breaks.

After the change:

screen-shot-2017-02-25-18-37-06

**Comment by renchap** [comment]
Nice find!
Do you think you can add a test for Turbolinks support? I think the simplest way would be to create a new controller with a Turbolinks layout, create a new pack that load Turbolinks, create 2 views in this controller, and have a test running `visit` on the first view, clicking on a link to the second, then to the first, and ensure that components are correctly loaded (and does not appear multiple times).

File: test/example_app/vendor/package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   "devDependencies": {
     "babel-core": "^6.22.1",
     "babel-loader": "^6.2.10",
+    "babel-preset-es2015": "^6.22.0",
```
Comment (renchap) [inline test/example_app/vendor/package.json]: This is not needed, `babel-preset-latest` already includes it

File: javascript/webpacker_react-npm-module/src/ujs.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
+const $ = (typeof window.jQuery !== 'undefined') && window.jQuery
```
Comment (renchap) [inline javascript/webpacker_react-npm-module/src/ujs.js]: Can you please add a comment saying this is adapted from rails-ujs?

**Comment by sevos** [comment]
Please note that I've had to make the init process explicit. IMHO it's better, importing a module should not have side effects.

**Comment by sevos** [comment]
I am done in this branch. Feel free to merge and lets discuss further API changes ;)

**Comment by renchap** [comment]
Merged 👍 

@sevos do you think it would be ok to call `initialize` on the first call to `register`? I would really like it to be called automatically to avoid any problems when users load multiple packs on the same page, so we need to ensure it is only called once in this case ; I am not sure of the behaviour of `this.eventsRegistered` when using multiple packs). Maybe we should replace this check by looking at the event listeners on `document` instead?

Feel free to send another PR with this change.

**Pull request merged**